### PR TITLE
Add reflecting boundary conditions for Svärd-Kalisch equations

### PR DESCRIPTION
**Pull request by JoshuaLampert**
In the case of $\alpha = \gamma = 0$ reflecting boundary conditions for the Svärd-Kalisch equations satisfy an entropy condition. This PR adds this possibility.

I used an LU factorization instead of a Cholesky decomposition for the system matrix for now because it was a bit simpler to implement and I was not 100% sure if the Dirichlet operator matrix $(\eta + D) - P_DD_1\hat{\beta}D_1$ is always guaranteed to be SPD (or rather the `[2:(end - 1), (2:end - 1)]` block of it). I would need to check that.

Regarding the treatment of the boundary conditions: We apply a Dirichlet operator in the equation for the velocity and explicitly set the boundary conditions `v_1 = v_N = 0` enforcing the boundary condition strongly. As I understand, for the mass equation no special treatment  of the boundary condition is necessary as we don't have any elliptic operator and a SAT term would simply be zero as the flux vanishes on the boundary (because the velocity vanishes on the boundary). Correct me if I'm wrong, @ranocha.

Convergence study looks a bit odd sometimes (e.g., convergence in $v$ for p = 6), but seems mostly ok:
<details>
<summary>p = 2</summary>

```julia
julia> convergence_test("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl", 4, N = 16, accuracy_order = 2);
[...]
####################################################################################################
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   3.82e-01  -         16   2.20e-02  -         16   0.00e+00  -         
32   1.27e-01  1.59      32   4.82e-03  2.19      32   0.00e+00  NaN       
64   4.44e-02  1.52      64   1.15e-03  2.07      64   0.00e+00  NaN       
128  1.58e-02  1.48      128  2.82e-04  2.03      128  0.00e+00  NaN       

mean           1.53      mean           2.10      mean           NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   1.77e+00  -         16   4.06e-02  -         16   0.00e+00  -         
32   8.73e-01  1.02      32   9.30e-03  2.13      32   0.00e+00  NaN       
64   4.33e-01  1.01      64   2.24e-03  2.05      64   0.00e+00  NaN       
128  2.15e-01  1.01      128  5.50e-04  2.03      128  0.00e+00  NaN       

mean           1.01      mean           2.07      mean           NaN       
----------------------------------------------------------------------------------------------------
```

</details>

<details>
<summary>p = 4</summary>

```julia
julia> convergence_test("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl", 4, N = 16, accuracy_order = 4);
[...]
####################################################################################################
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   1.62e-01  -         16   2.73e-03  -         16   0.00e+00  -         
32   1.62e-02  3.32      32   2.20e-04  3.64      32   0.00e+00  NaN       
64   1.74e-03  3.21      64   3.04e-05  2.85      64   0.00e+00  NaN       
128  2.21e-04  2.98      128  4.11e-06  2.89      128  0.00e+00  NaN       

mean           3.17      mean           3.12      mean           NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   5.99e-01  -         16   5.12e-03  -         16   0.00e+00  -         
32   5.68e-02  3.40      32   5.87e-04  3.12      32   0.00e+00  NaN       
64   6.22e-03  3.19      64   7.38e-05  2.99      64   0.00e+00  NaN       
128  1.55e-03  2.00      128  9.30e-06  2.99      128  0.00e+00  NaN       

mean           2.87      mean           3.03      mean           NaN       
----------------------------------------------------------------------------------------------------
```

</details>

<details>
<summary>p = 6</summary>

```julia
julia> convergence_test("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl", 4, N = 16, accuracy_order = 6);
####################################################################################################
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   7.65e-01  -         16   1.55e-02  -         16   0.00e+00  -         
32   6.70e-02  3.51      32   4.49e-05  8.43      32   0.00e+00  NaN       
64   3.97e-03  4.08      64   2.28e-05  0.98      64   0.00e+00  NaN       
128  3.06e-04  3.70      128  1.52e-06  3.91      128  0.00e+00  NaN       

mean           3.76      mean           4.44      mean           NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   2.63e+00  -         16   4.05e-02  -         16   0.00e+00  -         
32   3.68e-01  2.84      32   1.30e-04  8.28      32   0.00e+00  NaN       
64   2.72e-02  3.76      64   4.13e-05  1.65      64   0.00e+00  NaN       
128  2.37e-03  3.52      128  2.43e-06  4.09      128  0.00e+00  NaN       

mean           3.37      mean           4.68      mean           NaN       
----------------------------------------------------------------------------------------------------
```

</details>

TODO:
- [x] Upwind operators

### CODE REVIEW

File: src/equations/bbm_bbm_1d.jl [hunk 1]
```diff
@@ -338,8 +338,7 @@ function create_cache(mesh, equations::BBMBBMEquations1D{BathymetryFlat},
 
     # homogeneous Neumann boundary conditions
     if solver.D1 isa DerivativeOperator ||
-       solver.D1 isa UniformCoupledOperator ||
-       solver.D1 isa FourierDerivativeOperator
+       solver.D1 isa UniformCoupledOperator
```
Comment (JoshuaLampert) [inline src/equations/bbm_bbm_1d.jl:341]: This (and the other changes in this file) are unrelated, but I think we shouldn't allow a periodic `FourierDerivativeOperator` for solving a problem with reflecting boundary conditions.

File: src/equations/svaerd_kalisch_1d.jl [hunk 2]
```diff
@@ -185,15 +258,15 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
         minus_MD1betaD1 = D1mat' * Diagonal(M_beta) * D1mat
         system_matrix = let cache = (; M_h, minus_MD1betaD1)
             assemble_system_matrix!(cache, h,
-                                    D1, D1mat, equations)
```
Comment (JoshuaLampert) [inline src/equations/svaerd_kalisch_1d.jl:188]: This argument was unused.

File: src/equations/svaerd_kalisch_1d.jl [hunk 3]
```diff
@@ -160,7 +233,6 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
     alpha_eta_x_x = zero(h)
     y_x = zero(h)
     v_y_x = zero(h)
-    yv_x = zero(h)
```
Comment (JoshuaLampert) [inline src/equations/svaerd_kalisch_1d.jl:163]: This was not needed.

**Comment by JoshuaLampert** [comment]
Some test values change significantly between different operating systems. [In this example](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/actions/runs/12222702421/job/34093312538?pr=166#step:7:8828) we would need an absolute tolerance of ~1.0e-4 to let the MacOS test pass using the test values I get locally, while CI is successful for Linux and Windows. Do you know why these tests are particularly sensitive to rounding errors @ranocha? The condition number of the system matrix doesn't seem to be too high:
```julia
julia> include("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl");
[...]
julia> cond(Matrix(DispersiveShallowWater.assemble_system_matrix!(semi.cache, ones(N), equations)))
107031.29687711073
```
What do you think is the best way to deal with this? I am a bit hesitant to just put such a high absolute tolerance in the test.

File: src/equations/svaerd_kalisch_1d.jl [hunk 4]
```diff
@@ -140,6 +140,79 @@ function source_terms_manufactured(q, x, t, equations::SvaerdKalischEquations1D)
...
+
+    # Floating point errors accumulate a bit and the system matrix
+    # is not necessarily perfectly symmetric but only up to
+    # round-off errors. We wrap it here to avoid issues with the
+    # factorization.
+    return Symmetric(Diagonal(M_h) + minus_MD1betaD1)
+end
+
+# For reflecting boundary conditions
+function assemble_system_matrix!(cache, h,
```
Comment (JoshuaLampert) [inline src/equations/svaerd_kalisch_1d.jl:210]: This function is not actually mutating, but I would like to use the same function name for the same functionality and then dispatch on the arguments. Hence, the bang in the name. Is this fine with you, @ranocha?

Comment (ranocha) [reply src/equations/svaerd_kalisch_1d.jl:210]: 👍

File: src/equations/svaerd_kalisch_1d.jl [hunk 5]
```diff
@@ -211,18 +283,54 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
...
-    return Symmetric(Diagonal(M_h) + minus_MD1betaD1)
+# Reflecting boundary conditions assume alpha = gamma = 0
+function create_cache(mesh, equations::SvaerdKalischEquations1D,
+                      solver, initial_condition,
+                      ::BoundaryConditionReflecting,
+                      RealT, uEltype)
+    if !isapprox(equations.alpha, 0, atol = 1e-14) ||
+       !isapprox(equations.gamma, 0, atol = 1e-14)
+        throw(ArgumentError("Reflecting boundary conditions for Svärd-Kalisch equations only implemented for alpha = gamma = 0"))
+    end
```
Comment (ranocha) [inline src/equations/svaerd_kalisch_1d.jl:294]: Wouldn't it be simpler to test

```suggestion
    if !iszero(equations.alpha) || !iszero(equations.gamma,)
        throw(ArgumentError("Reflecting boundary conditions for Svärd-Kalisch equations only implemented for alpha = gamma = 0"))
    end
```

The current test disallows `eps(Float32)` deviations for 32-bit computations anyway.

File: src/equations/svaerd_kalisch_1d.jl [hunk 6]
```diff
@@ -211,18 +283,54 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
...
+        D[i] = still_waterdepth(initial_condition(x[i], 0.0, equations, mesh), equations)
+    end
+    h = ones(RealT, N)
+    hv = zero(h)
+    b = zero(h)
+    eta_x = zero(h)
+    v_x = zero(h)
+    h_v_x = zero(h)
+    hv2_x = zero(h)
+    beta_hat = equations.beta * D .^ 3
```
Comment (ranocha) [inline src/equations/svaerd_kalisch_1d.jl:310]:
```suggestion
    beta_hat = equations.beta .* D .^ 3
```

File: src/equations/svaerd_kalisch_1d.jl [hunk 7]
```diff
@@ -326,6 +434,52 @@ function rhs!(dq, q, t, mesh, equations::SvaerdKalischEquations1D,
     return nothing
 end
 
+function rhs!(dq, q, t, mesh, equations::SvaerdKalischEquations1D,
+              initial_condition, ::BoundaryConditionReflecting, source_terms,
+              solver, cache)
+    (; D, h, hv, b, eta_x, v_x, h_v_x, hv2_x, tmp1, D1_central) = cache
```
Comment (ranocha) [inline src/equations/svaerd_kalisch_1d.jl:440]:
```suggestion
    # When constructing the `cache`, we assert that alpha and gamma are zero.
    # We use this explicitly in the code below.
    (; D, h, hv, b, eta_x, v_x, h_v_x, hv2_x, tmp1, D1_central) = cache
```

File: src/equations/svaerd_kalisch_1d.jl [hunk 8]
```diff
@@ -326,6 +434,52 @@ function rhs!(dq, q, t, mesh, equations::SvaerdKalischEquations1D,
...
+        @.. tmp1 = hv * v
+        mul!(hv2_x, D1_central, tmp1)
+        @.. dv = -(0.5 * (hv2_x + hv * v_x - v * h_v_x) +
+                   g * h * eta_x)
+    end
+
+    # no split form
+    #     dv[:] = -(D1_central * (hv .* v) - v .* (D1_central * hv)+
+    #               g * h .* eta_x)
+
```
Comment (ranocha) [inline src/equations/svaerd_kalisch_1d.jl:469]:
```suggestion
```
?

**Comment by ranocha** [comment]
> Some test values change significantly between different operating systems. [In this example](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/actions/runs/12222702421/job/34093312538?pr=166#step:7:8828) we would need an absolute tolerance of ~1.0e-4 to let the MacOS test pass using the test values I get locally, while CI is successful for Linux and Windows. 

It also fails on Linux with Julia 1.11, see https://github.com/JoshuaLampert/DispersiveShallowWater.jl/actions/runs/12222702421/job/34093312682?pr=166

> Do you know why these tests are particularly sensitive to rounding errors @ranocha? The condition number of the system matrix doesn't seem to be too high:

This is for the central SBP operators, isn't it? We know that these methods are not particularly stable for elliptic solves.

> What do you think is the best way to deal with this? I am a bit hesitant to just put such a high absolute tolerance in the test. 

I agree. However, if upwind is fine, we may need to accept it 🤷 Does it help if you set stricter tolerances of the time integration methods?

> Edit: Could it be that using a Cholesky decomposition helps stabilizing the method?

I am not sure

**Comment by JoshuaLampert** [comment]
> It also fails on Linux with Julia 1.11, see [JoshuaLampert/DispersiveShallowWater.jl/actions/runs/12222702421/job/34093312682?pr=166](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/actions/runs/12222702421/job/34093312682?pr=166)

Yes, but it's not as severe. A tolerance of ~1e-7 would be enough. But yes, it is still quite high.

> This is for the central SBP operators, isn't it? We know that these methods are not particularly stable for elliptic solves.

The tests fail for both central and upwind operators. For the central operators we would a tolerance of 1e-4 and for the upwind operators 1e-5. The condition number I posted is for central operators.

> I agree. However, if upwind is fine, we may need to accept it 🤷 Does it help if you set stricter tolerances of the time integration methods?

Good idea. I'll try.

> > Edit: Could it be that using a Cholesky decomposition helps stabilizing the method?
>
> I am not sure

Maybe it's worth a try. A Cholesky decomposition should be desirable anyway. And I think the matrix is SPD. I'll give it a try.

**Comment by JoshuaLampert** [comment]
> > I agree. However, if upwind is fine, we may need to accept it 🤷 Does it help if you set stricter tolerances of the time integration methods?
>
> Good idea. I'll try.

Seems like, it indeed really helps. With `abstol` and `reltol` equal to 1e-12, we only need test tolerances of 1e-11, which is acceptable.

**Comment by ranocha** [comment]
Sounds good 👍

**Comment by JoshuaLampert** [comment]
I have now switched from an LU decomposition to a Cholesky decomposition in [40468f3](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/pull/166/commits/40468f3650fb5436f90a3abfbad23ab248c44ffd), but the errors are too high. So it seems there is some bug somewhere. However, the `system_matrix` seems to be symmetric and positive definite. Do you have an idea, where things go wrong, @ranocha?

File: src/equations/svaerd_kalisch_1d.jl [hunk 9]
```diff
@@ -255,17 +265,17 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
...
             assemble_system_matrix!(cache, h,
-                                    D1, equations)
+                                    equations, boundary_conditions)
         end
     elseif D1 isa PeriodicUpwindOperators
         D1_central = D1.central
         D1mat_minus = sparse(D1.minus)
         minus_MD1betaD1 = D1mat_minus' * Diagonal(M_beta) * D1mat_minus
-        system_matrix = let cache = (; M_h, minus_MD1betaD1)
+        system_matrix = let cache = (; D1, M_h, minus_MD1betaD1)
```
Comment (ranocha) [inline src/equations/svaerd_kalisch_1d.jl:276]: Same here?

File: src/equations/svaerd_kalisch_1d.jl [hunk 10]
```diff
@@ -255,17 +265,17 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
         D1_central = D1
         D1mat = sparse(D1_central)
         minus_MD1betaD1 = D1mat' * Diagonal(M_beta) * D1mat
-        system_matrix = let cache = (; M_h, minus_MD1betaD1)
+        system_matrix = let cache = (; D1, M_h, minus_MD1betaD1)
             assemble_system_matrix!(cache, h,
```
[the remark this answers is not in the document]
Comment (JoshuaLampert) [reply src/equations/svaerd_kalisch_1d.jl:269]: What exactly do you mean by "this"? The `h` that is passed to `assemble_system_matrix!`? This is scaled to `M_h` within that function.

[the remark this answers is not in the document]
Comment (ranocha) [reply src/equations/svaerd_kalisch_1d.jl:269]: Right, sorry

**Comment by ranocha** [comment]
Did you perform a convergence test with the new implementation?

**Comment by JoshuaLampert** [comment]
```julia
julia> convergence_test("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl", 4, N = 16, accuracy_order = 6)
[...]
####################################################################################################
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   2.42e+00  -         16   1.39e-01  -         16   0.00e+00  -         
32   6.55e-01  1.89      32   2.30e-02  2.60      32   0.00e+00  NaN       
64   1.79e-01  1.87      64   4.38e-03  2.39      64   0.00e+00  NaN       
128  5.63e-02  1.67      128  9.58e-04  2.19      128  0.00e+00  NaN       

mean           1.81      mean           2.39      mean           NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   6.34e+00  -         16   3.53e-01  -         16   0.00e+00  -         
32   2.35e+00  1.43      32   4.31e-02  3.03      32   0.00e+00  NaN       
64   8.58e-01  1.45      64   6.95e-03  2.63      64   0.00e+00  NaN       
128  3.63e-01  1.24      128  1.36e-03  2.35      128  0.00e+00  NaN       

mean           1.38      mean           2.67      mean           NaN       
----------------------------------------------------------------------------------------------------
```

**Comment by ranocha** [comment]
I don't see any obvious error in the code at the moment. Could you maybe setup a simple example and check how the two versions (LU and Cholesky) behave when solving a single linear system?

**Comment by JoshuaLampert** [comment]
Do you mean something like this:
```julia
julia> include("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl");
[...]
julia> t = 0.1
0.1

julia> q = DispersiveShallowWater.compute_coefficients(initial_condition, t, semi)
([1.2214027581601699, 1.2213796755201909, 1.2213104284727068, 1.2211950196350434, 1.2210334533693017, 1.2208257357821906, 1.2205718747247982, 1.2202718797922947, 1.2199257623235684, 1.219533535400799  …  -1.219533535400799, -1.2199257623235684, -1.2202718797922947, -1.2205718747247982, -1.2208257357821906, -1.2210334533693017, -1.2211950196350434, -1.2213104284727068, -1.2213796755201909, -1.2214027581601699], [0.0, 1.3296421832613442e-5, 5.3184682202485554e-5, 0.0001196617657826978, 0.00021272264721833685, 0.00033236029141141103, 0.00047856565391971273, 0.0006513276814696079, 0.000850633312582725, 0.0010764674783165142  …  0.06004296379054353, 0.05348356952863884, 0.04689559306581195, 0.04027927587157583, 0.033634861490834986, 0.0269625955349242, 0.02026272567253702, 0.013535501620532575, 0.006781175134633048, 0.0], [7.0, 6.999848813690903, 6.9993952776209145, 6.9986394603584765, 6.997581476172813, 6.996221485016647, 6.994559692502023, 6.992596349869215, 6.9903317539487535, 6.987766247116534  …  6.987766247116534, 6.9903317539487535, 6.992596349869215, 6.994559692502023, 6.996221485016647, 6.997581476172813, 6.9986394603584765, 6.9993952776209145, 6.999848813690903, 7.0])

julia> dq = similar(q)
([9.7969281e-316, 9.70988004e-316, 9.7969281e-316, 9.70988004e-316, 6.0e-323, 6.518773660173e-311, 6.4e-323, 7.0620045826603e-311, 7.0e-323, 7.6052355051474e-311  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [9.7969281e-316, 9.8260313e-316, 9.7969281e-316, 9.8260313e-316, 2.172928870602e-311, 2.7161597930893e-311, 5.432314405525e-312, 1.0864623630396e-311, 1.6296932855267e-311, 5.180654e-318  …  4.243991582e-313, 0.0, 3.75e-321, 2.7585945283e-313, 0.0, 3.814e-321, 2.9707941074e-313, 0.0, 4.076e-321, 1.9097962119e-313], [1.5993478e-316, 9.8260349e-316, 9.8260349e-316, 0.0, 9.8260349e-316, 3.78818866e-316, 3.78818905e-316, 3.78818905e-316, 9.82603764e-316, 9.82603764e-316  …  2.53e-321, 2.100775833056e-312, 2.0825436e-316, 6.50493647688795e-310, 3.5e-323, 2.54e-321, 2.100775833056e-312, 2.7170591e-316, 6.50493647688795e-310, 3.5e-323])

julia> DispersiveShallowWater.rhs!(dq, q, t, mesh, equations, initial_condition, boundary_conditions, source_terms, solver, semi.cache)

julia> dq_lu = similar(q)
([8.487983164e-314, 4.14173848e-316, 4.243991582e-314, 7.2792943613299e-310, NaN, 4.14174006e-316, 4.243991582e-314, 8.487983164e-314, 8.4879831856e-314, 4.14174164e-316  …  NaN, 1.586e-321, 7.42420905e-316, 7.4196439e-316, NaN, NaN, NaN, NaN, NaN, NaN], [2.5e-323, 6.7765372e-316, 1.3190567232077176e-84, 4.0e-323, 0.0, 5.0e-324, 0.0, 1.69759663287e-313, 2.121995791e-314, 0.0  …  0.0, 6.96e-321, 6.50494391036804e-310, 4.92504477e-316, 0.0, 0.0, 3.28093665e-316, 6.94252987e-316, 4.03878433e-316, 9.8453521e-316], [6.50494391039334e-310, 4.1419547e-316, 4.1419547e-316, 4.1419547e-316, 1.497e-321, 5.0e-324, 0.0, 0.0, 1.7381080806e-314, 6.5049397079524e-310  …  4.0387843e-316, 3.8132271e-316, NaN, NaN, NaN, 3.00831513e-316, NaN, 4.14055865e-316, NaN, 4.38453854e-316])

julia> DispersiveShallowWater.rhs_lu!(dq_lu, q, t, mesh, equations, initial_condition, boundary_conditions, source_terms, solver, semi.cache)

julia> dq .- dq_lu
([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.0, 2.5885250037231663e-5, -1.62061519686071e-5, 3.576633358458717e-5, -1.0175667095673287e-5, 3.6540107521403726e-5, -1.0095704447866147e-5, 3.653070993839909e-5, -1.011872634096723e-5, 3.650777580948075e-5  …  -3.877105139398623e-5, 1.1622132469489566e-5, -3.8796828582446374e-5, 1.1602498280804951e-5, -3.880869896399214e-5, 1.1689689387396618e-5, -3.7999106218711276e-5, 1.8091870581942934e-5, -2.7143511910356424e-5, 0.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
```
There are differences of the order of 1e-5 in the velocity.

**Comment by JoshuaLampert** [comment]
Ohh, I found the bug ([f017619](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/pull/166/commits/f01761985db1fee95551c9a83655e24d79e6c405) fixes this). `scale_by_mass_matrix!` was in used `solve_system_matrix!` for a smaller vector than the size of the mass matrix. Let me clean up this PR. Then this should be fine.

**Comment by JoshuaLampert** [comment]
> > `scale_by_mass_matrix!` was used in `solve_system_matrix!` for a smaller vector than the size of the mass matrix. Why doesn't this trigger the boundscheck in [ranocha/SummationByPartsOperators.jl@00164c7/src/matrix_operators.jl#L55](https://github.com/ranocha/SummationByPartsOperators.jl/blob/00164c715f6df4f1ea7297d28e2a39ade3334223/src/matrix_operators.jl#L55)?
> 
> That's a good question...

I opened an issue: https://github.com/ranocha/SummationByPartsOperators.jl/issues/310

**Comment by JoshuaLampert** [comment]
This is ready for a review @ranocha.

File: examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+summary_callback = SummaryCallback()
+analysis_callback = AnalysisCallback(semi; interval = 100,
+                                     extra_analysis_errors = (:conservation_error,),
+                                     extra_analysis_integrals = (waterheight_total,
+                                                                 entropy_modified))
+callbacks = CallbackSet(analysis_callback, summary_callback)
+
+saveat = range(tspan..., length = 100)
+sol = solve(ode, Tsit5(), abstol = 1e-12, reltol = 1e-12,
+            save_everystep = false, callback = callbacks, saveat = saveat)
```
[the remark this answers is not in the document]
Comment (JoshuaLampert) [reply examples/svaerd_kalisch_1d/svaerd_kalisch_1d_basic_reflecting.jl]: Done in [7358093](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/pull/166/commits/7358093cff699397ac2583ffaad4cde2a9021f0f).

File: src/equations/svaerd_kalisch_1d.jl [hunk 12]
```diff
@@ -211,18 +293,58 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
...
-    (; M_h, minus_MD1betaD1) = cache
-
-    @.. M_h = h
+# Reflecting boundary conditions assume alpha = gamma = 0
+function create_cache(mesh, equations::SvaerdKalischEquations1D,
+                      solver, initial_condition,
+                      boundary_conditions::BoundaryConditionReflecting,
+                      RealT, uEltype)
+    if !iszero(equations.alpha) || !iszero(equations.gamma)
+        throw(ArgumentError("Reflecting boundary conditions for Svärd-Kalisch equations only implemented for alpha = gamma = 0"))
```
[the remark this answers is not in the document]
Comment (JoshuaLampert) [reply src/equations/svaerd_kalisch_1d.jl:302]: Done in [7092265](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/pull/166/commits/7092265877d86b41c3cffbb346a03b269510c7e9).

**Review by ranocha — APPROVED** [review]
Thanks!
